src/ctorch.cpp

```
@@ -32,25 +50,23 @@ constexpr auto get_libtorch_dtype(torch_data_t dtype) {
   case torch_kInt64:
     return torch::kInt64;
   case torch_kFloat16:
-    std::cerr << "[WARNING]: float16 not supported in Fortran" << std::endl;
+    ctorch_error("float16 not supported in Fortran");
     // See https://gcc.gnu.org/onlinedocs/gfortran/ISO_005fFORTRAN_005fENV.html
-    exit(EXIT_FAILURE);
   case torch_kFloat32:
     return torch::kFloat32;
   case torch_kFloat64:
     return torch::kFloat64;
   default:
-    std::cerr << "[WARNING]: unknown data type, setting to torch_kFloat32" << std::endl;
+    ctorch_error("unknown data type, setting to torch_kFloat32", true);
```

I find calling `ctorch_error` to raise a warning a bit unclear. Given that the function has different implementations for the error and warning cases, would it not make more sense to just have separate `ctorch_error` and `ctorch_warning` functions? That'd also make it clearer because currently the `cleanup` function gets silently ignored if you are just raising a warning.

---

Yes, that's fair and I did wonder about doing this as the function developed.
Should be covered in https://github.com/Cambridge-ICCS/FTorch/pull/347/commits/479d39ec0e82b938b16770f2c0193e7deb9e4be8